src/pycapsule.rs

```diff
@@ -0,0 +1,325 @@
+///
+///     let capsule = PyCapsule::new(py, foo, name.as_ref())?;
+///
+///     let module = PyModule::import(py, "builtins")?;
+///     module.add("capsule", capsule)?;
+///
+///     let cap: &Foo = unsafe { PyCapsule::import(py, name.as_ref())? };
+///     assert_eq!(cap.val, 123);
+///     Ok(())
+///  });
+///  assert!(r.is_ok());
+/// ```
+#[repr(transparent)]
+pub struct PyCapsule(PyAny);
+
+pyobject_native_type_core!(PyCapsule, ffi::PyCapsule_Type, #checkfunction=ffi::PyCapsule_CheckExact);
+
+impl PyCapsule {
+    /// Constructs a new capsule whose contents are `value`, associated with `name`.
+    ///
+    /// # Notes
+    ///
+    /// An attempt to add a zero sized value will panic.
+    pub fn new<'py, T: 'static + Send>(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+    ) -> PyResult<&'py Self> {
+        Self::new_with_destructor(py, value, name, |_, _| {})
+    }
+
+    /// Constructs a new capsule whose contents are `value`, associated with `name`.
+    ///
+    /// Also provides a destructor: when the `PyCapsule` is destroyed, it will be passed the original object,
+    /// as well as `*mut c_void` which will point to the capsule's context, if any.
+    ///
+    /// # Notes
+    ///
+    /// An attempt to add a zero sized value will panic.
+    pub fn new_with_destructor<'py, T: 'static + Send, F: FnOnce(T, *mut c_void)>(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+        destructor: F,
+    ) -> PyResult<&'py Self> {
+        assert!(
+            size_of::<T>() > 0,
+            "Zero sized objects not allowed in capsule."
+        );
+        let val = Box::new(CapsuleContents { value, destructor });
+
+        let cap_ptr = unsafe {
+            ffi::PyCapsule_New(
+                Box::into_raw(val) as *mut c_void,
+                name.as_ptr(),
+                Some(capsule_destructor::<T, F>),
+            )
+        };
+        unsafe { py.from_owned_ptr_or_err(cap_ptr) }
+    }
+
+    /// Imports an existing capsule.
+    ///
+    /// The `name` should match the path to the module attribute exactly in the form
+    /// of `module.attribute`, which should be the same as the name within the capsule.
+    ///
+    /// # Safety
+    ///
+    /// It must be known that this capsule's pointer is to an item of type `T`.
+    pub unsafe fn import<'py, T>(py: Python<'py>, name: &CStr) -> PyResult<&'py T> {
+        let ptr = ffi::PyCapsule_Import(name.as_ptr(), false as c_int);
+        if ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(&*(ptr as *const T))
+        }
+    }
+
+    /// Sets the context pointer in the capsule to `T`.
+    ///
+    /// # Notes
+    ///
+    /// Context is not destructed like the value of the capsule is by `destructor`. Therefore,
+    /// it's likely this value will be leaked. If set, it's up to the user to either ignore this
+    /// side effect or figure another (unsafe) method of clean up.
```

Comment:
This seems a bit outdated now that the value is passed into the destructor. Suggestion:

```
The default destructor does not drop the context value. If you need to ensure it is not leaked, use `new_with_destructor`, let it call `Box::from_raw()` with the correct type, and drop it.

Finally, if `set_context` is called twice in a row, the previous value is always leaked.
```

It seems that it would be better to treat the context like the value, i.e. another type parameter for `new` and the destructor. That would also allow dropping the previous context properly on `set_context`.

However, this depends on how context is supposed to be used (which I don't know either): if the context is supposed to be set by the other side (which might well be C), this would not help, and rather be detrimental since the destructor would try to drop something non-Rust.

It's all a bit iffy. I wonder if we should instead simply have the API operate on raw pointers. People can still call `Box::into_raw` for Rust objects, and it would be more symmetric between here and the destructor. @davidhewitt ?

Reply:
Yes, I was beginning to wonder whether the `set_context` should just take `*mut void` too. If we have an example which demonstrates `into_raw` (and `from_raw` in the destructor) then there's something for them to copy.

Because anything can mutate the capsule using the raw FFI calls, I'm leaning away from us trying to do anything too smart - at some point it's better if the `unsafe` aspects of this are acknowledged by the user rather than just papered over by a wrapper from PyO3 which has invariants we can't guarantee are maintained.

Reply:
(`get_context` may need an equivalent change to `*mut void` too, I guess?)

Reply:
Agreed, let's do it that way.